pkg/controller/restore_controller.go

```diff
@@ -462,43 +461,97 @@ func mostRecentCompletedBackup(backups []*api.Backup) *api.Backup {
 	return nil
 }
 
-func (c *restoreController) fetchBackup(objectStore cloudprovider.ObjectStore, name string) (*api.Backup, error) {
-	backup, err := c.backupLister.Backups(c.namespace).Get(name)
-	if err == nil {
-		return backup, nil
+// fetchBackupInfo checks the object storage for a backup that matches the given name. If it doesn't
+// find it, it tries to retrieve it from one of the backup storage locations.
+func (c *restoreController) fetchBackupInfo(backupName string, pluginManager plugin.Manager) (backupInfo, error) {
+	var info backupInfo
+	var err error
+	info.backup, err = c.backupLister.Backups(c.namespace).Get(backupName)
+	if err != nil {
+		return c.fetchFallBack(backupName, pluginManager)
 	}
 
-	if !apierrors.IsNotFound(err) {
-		return nil, errors.WithStack(err)
+	if info.backup.Spec.StorageLocation == "" {
+		info.backup.Spec.StorageLocation = c.defaultBackupLocation
 	}
 
-	logContext := c.logger.WithField("backupName", name)
+	location, err := c.backupLocationLister.BackupStorageLocations(c.namespace).Get(info.backup.Spec.StorageLocation)
+	if err != nil {
+		return backupInfo{}, err
+	}
 
-	logContext.Debug("Backup not found in backupLister, checking object storage directly")
-	backup, err = c.getBackup(objectStore, c.bucket, name)
+	info.objectStore, err = getObjectStoreForLocation(location, pluginManager)
 	if err != nil {
-		return nil, err
+		return backupInfo{}, errors.Wrap(err, "error initializing object store")
 	}
 
-	// ResourceVersion needs to be cleared in order to create the object in the API
-	backup.ResourceVersion = ""
-	// Clear out the namespace too, just in case
-	backup.Namespace = ""
+	return info, nil
+}
 
-	created, createErr := c.backupClient.Backups(c.namespace).Create(backup)
-	if createErr != nil {
-		logContext.WithError(errors.WithStack(createErr)).Error("Unable to create API object for Backup")
-	} else {
-		backup = created
+// fetchFallBack checks each backup storage location, starting with the default,
+// looking for a backup that matches the given backup name.
+func (c *restoreController) fetchFallBack(backupName string, pluginManager plugin.Manager) (backupInfo, error) {
+	var info backupInfo
+	var err error
+
+	logContext := c.logger.WithField("backupName", backupName)
+	logContext.Debug("Backup not found in backupLister, checking each backup location directly, starting with default...")
+
+	location, err := c.backupLocationLister.BackupStorageLocations(c.namespace).Get(c.defaultBackupLocation)
```

Comment:
Instead of needing a separate largely copy-pasted block of code to check the default location first, how about we get the full list of all BackupStorageLocations (per https://github.com/heptio/ark/pull/766/files#diff-7d210f9ced78ff8be1ba497bf49e1a40R512), then construct an ordered slice of locations to iterate through, where the first one is the default location, and the rest are the remainder, in any order